src/goto-analyzer/get_function_pointer_values.cpp

```diff
@@ -0,0 +1,113 @@
+/*******************************************************************\
+
+Module: Show function pointer values
+
+Author: Diffblue Ltd, 2020
+
+\*******************************************************************/
+
+#include "get_function_pointer_values.h"
+
+#include <util/message.h>
+#include <util/namespace.h>
+
+#include <goto-programs/goto_model.h>
+#include <goto-programs/restrict_function_pointers.h>
+
+#include <analyses/ai.h>
+#include <analyses/variable-sensitivity/abstract_enviroment.h>
+#include <analyses/variable-sensitivity/value_set_abstract_object.h>
+#include <analyses/variable-sensitivity/variable_sensitivity_domain.h>
+
+#include <regex>
+#include <string>
+
+void get_function_pointer_values(
+  const goto_modelt &goto_model,
+  const ai_baset &ai,
+  const optionst &options,
+  message_handlert &message_handler,
+  std::string filename)
+{
+  function_pointer_restrictionst::restrictionst map;
+  namespacet ns(goto_model.symbol_table);
+
+  messaget m(message_handler);
+  m.status() << "Getting function pointer targets at specific callsites"
+             << messaget::eom;
+
+  for(auto const &function_entry : goto_model.goto_functions.function_map)
+  {
+    for_each_instruction(
+      function_entry.second, [&](ai_baset::locationt instruction_location) {
+        auto is_assign = instruction_location->is_assign();
+        if(is_assign)
+        {
+          auto assign_stmt = instruction_location->get_assign();
+          auto lhs = assign_stmt.lhs();
+          std::string identifier =
+            id2string(to_symbol_expr(lhs).get_identifier());
+          // <name of function>.function_pointer_call.<integer>
+          std::regex name(R"(\w+\.function_pointer_call\.\d+)");
+          if(std::regex_match(identifier, name))
```

Comment:
This whole block represents a function that can be easily pulled out that goes from identifier to boolean of whether it is a relevant assignment